ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java

```diff
@@ -0,0 +1,777 @@
+ */
+package com.github.ambry.router;
+
+import com.github.ambry.clustermap.ClusterMap;
+import com.github.ambry.clustermap.ReplicaId;
+import com.github.ambry.commons.BlobId;
+import com.github.ambry.commons.BlobIdFactory;
+import com.github.ambry.commons.ResponseHandler;
+import com.github.ambry.commons.ServerErrorCode;
+import com.github.ambry.config.RouterConfig;
+import com.github.ambry.messageformat.BlobData;
+import com.github.ambry.messageformat.BlobType;
+import com.github.ambry.messageformat.MessageFormatException;
+import com.github.ambry.messageformat.MessageFormatFlags;
+import com.github.ambry.messageformat.MessageFormatRecord;
+import com.github.ambry.messageformat.MetadataContentSerDe;
+import com.github.ambry.network.Port;
+import com.github.ambry.network.RequestInfo;
+import com.github.ambry.network.ResponseInfo;
+import com.github.ambry.protocol.GetRequest;
+import com.github.ambry.protocol.GetResponse;
+import com.github.ambry.protocol.RequestOrResponse;
+import com.github.ambry.store.StoreKey;
+import com.github.ambry.utils.ByteBufferInputStream;
+import com.github.ambry.utils.Time;
+import com.github.ambry.utils.Utils;
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * GetBlobOperation class is responsible for maintaining the state associated with a GetBlob operation, and completing
+ * it. It would help to look at this class in conjunction with {@link PutOperation}.
+ *
+ * Blobs that need to be fetched are either simple (those that consist of only a single chunk) or composite (those
+ * that consists of one metadata chunk and two or more data chunks). Whether a blob is simple or composite can only be
+ * determined after the first chunk is fetched, based on whether the first chunk turns out to be a metadata chunk or
+ * a data chunk.
+ *
+ * As soon as the first data chunk is fetched (which could be the first chunk in the case of simple blobs and the
+ * second chunk fetched in the case of composite blobs), the operation callback is invoked (and the future is marked
+ * as done) so that the caller can start reading in data. The rest of the chunks are asynchronously fetched and
+ * buffered up to the maximum that can be buffered. When fetched chunks are consumed by the caller, subsequent chunks
+ * become eligible to be fetched.
+ *
+ * This class implements {@link ReadableStreamChannel} and passes itself as the result of this operation.
+ */
+class GetBlobOperation extends GetOperation<ReadableStreamChannel> implements ReadableStreamChannel {
```

Comment:
Not sure how you feel about this but `GetOperation` and `ReadableStreamChannel` are two very different kinds of functionality. It is like saying an operation is also a channel. 

If you think this is the simplest and easiest way, I'm ok with that but I would have preferred an inner class implementing `ReadableStreamChannel`.